File: src/parser.ts
import { Token, LineSpecError } from './lexer';
import {
  TestSpec,
  ReceiveStatement,
  RespondStatement,
  ExpectStatement,
  ExpectHttpStatement,
  ExpectReadMysqlStatement,
  ExpectWriteMysqlStatement,
  ExpectWritePostgresqlStatement,
  ExpectEventStatement,
  ExpectNotStatement,
  ExpectNotHttpStatement,
  ExpectNotWriteMysqlStatement,
  ExpectNotWritePostgresqlStatement,
  ExpectNotEventStatement,
  VerifyRule,
} from './types';
import * as path from 'path';

function peek(tokens: Token[], pos: number): Token | undefined {
  return tokens[pos];
}

function consume(tokens: Token[], pos: { value: number }): Token {
  const token = tokens[pos.value];
  if (!token) {
    throw new LineSpecError('Unexpected end of input');
  }
  pos.value++;
  return token;
}

function expectType(tokens: Token[], pos: { value: number }, type: Token['type']): Token {
  const token = consume(tokens, pos);
  if (token.type !== type) {
    throw new LineSpecError(`Expected ${type} but got ${token.type}`, token.line);
  }
  return token;
}

function parseVerifyRule(value: string, line: number): VerifyRule {
  // Format: query CONTAINS 'string' or query NOT_CONTAINS 'string' or query MATCHES /regex/
  const containsMatch = value.match(/^query\s+CONTAINS\s+['"](.+?)['"]$/i);
  if (containsMatch) {
    return { type: 'CONTAINS', pattern: containsMatch[1] };
  }
  
  const notContainsMatch = value.match(/^query\s+NOT_CONTAINS\s+['"](.+?)['"]$/i);
  if (notContainsMatch) {
    return { type: 'NOT_CONTAINS', pattern: notContainsMatch[1] };
  }
  
  const matchesMatch = value.match(/^query\s+MATCHES\s\/(.+?)\/$/i);
  if (matchesMatch) {
    return { type: 'MATCHES', pattern: matchesMatch[1] };
  }
  
  throw new LineSpecError(`Invalid VERIFY format: ${value}. Expected: VERIFY query CONTAINS 'string', VERIFY query NOT_CONTAINS 'string', or VERIFY query MATCHES /regex/`, line);
}

function parseExpectNotChannel(
  value: string,
  line: number
): Partial<ExpectNotStatement> {
  const eventMatch = value.match(/^(EVENT|MESSAGE):(.+)$/i);
  if (eventMatch) {
    const expectNot: ExpectNotEventStatement = {
      channel: 'EVENT',
      topic: eventMatch[2],
    };
    return expectNot;
  }

  const firstSpace = value.indexOf(' ');
  if (firstSpace === -1) {
    throw new LineSpecError(`Invalid EXPECT NOT channel format: ${value}`, line);
  }

  const channelPart = value.substring(0, firstSpace).toUpperCase();
  const rest = value.substring(firstSpace + 1);

  const httpMatch = channelPart.match(/^HTTP:(\w+)$/i);
  if (httpMatch) {
    const expectNot: ExpectNotHttpStatement = {
      channel: 'HTTP',
      method: httpMatch[1].toUpperCase(),
      url: rest,
    };
    return expectNot;
  }

  const writeMysqlMatch = channelPart.match(/^WRITE:MYSQL$/i);
  if (writeMysqlMatch) {
    const operationMatch = rest.match(/^(INSERT|UPDATE|DELETE)\s+(.+)$/i);
    let operation: 'INSERT' | 'UPDATE' | 'DELETE' | undefined;
    let table: string;
    
    if (operationMatch) {
      operation = operationMatch[1].toUpperCase() as 'INSERT' | 'UPDATE' | 'DELETE';
      table = operationMatch[2];
    } else {
      table = rest;
    }
    
    const expectNot: ExpectNotWriteMysqlStatement = {
      channel: 'WRITE_MYSQL',
      table,
      operation,
    };
    return expectNot;
  }

  if (/^WRITE:POSTGRESQL$/i.test(channelPart)) {
    const expectNot: ExpectNotWritePostgresqlStatement = {
      channel: 'WRITE_POSTGRESQL',
      table: rest,
    };
    return expectNot;
  }

  throw new LineSpecError(`Unrecognized EXPECT NOT channel: ${channelPart}`, line);
}

function parseExpectChannel(
  value: string,
  line: number
): Partial<ExpectStatement> {
  const eventMatch = value.match(/^(EVENT|MESSAGE):(.+)$/i);
  if (eventMatch) {
    const expect: ExpectEventStatement = {
      channel: 'EVENT',
      topic: eventMatch[2],
      withFile: '',
    };
    return expect;
  }

  const firstSpace = value.indexOf(' ');
  if (firstSpace === -1) {
    throw new LineSpecError(`Invalid EXPECT channel format: ${value}`, line);
  }

  const channelPart = value.substring(0, firstSpace).toUpperCase();
  const rest = value.substring(firstSpace + 1);

  const httpMatch = channelPart.match(/^HTTP:(\w+)$/i);
  if (httpMatch) {
    const expect: ExpectHttpStatement = {
      channel: 'HTTP',
      method: httpMatch[1].toUpperCase(),
      url: rest,
      returnsFile: '',
    };
    return expect;
  }

  const writeMysqlMatch = channelPart.match(/^WRITE:MYSQL$/i);
  if (writeMysqlMatch) {
    // Parse optional operation type from rest (e.g., "DELETE users" or just "users")
    const operationMatch = rest.match(/^(INSERT|UPDATE|DELETE)\s+(.+)$/i);
    let operation: 'INSERT' | 'UPDATE' | 'DELETE' | undefined;
    let table: string;
    
    if (operationMatch) {
      operation = operationMatch[1].toUpperCase() as 'INSERT' | 'UPDATE' | 'DELETE';
      table = operationMatch[2];
    } else {
      table = rest;
    }
    
    const expect: ExpectWriteMysqlStatement = {
      channel: 'WRITE_MYSQL',
      table,
      operation,
      withFile: '',
      returnsFile: '',
      transactional: true, // Default to transactional
    };
    return expect;
  }

  if (/^READ:MYSQL$/i.test(channelPart)) {
    const expect: ExpectReadMysqlStatement = {
      channel: 'READ_MYSQL',
      table: rest,
      returnsFile: '',
    };
    return expect;
  }

  if (/^WRITE:POSTGRESQL$/i.test(channelPart)) {
    const expect: ExpectWritePostgresqlStatement = {
      channel: 'WRITE_POSTGRESQL',
      table: rest,
      withFile: '',
      returnsFile: '',
    };
    return expect;
  }

  throw new LineSpecError(`Unrecognized EXPECT channel: ${channelPart}`, line);
}

export { LineSpecError };

export function parse(tokens: Token[], filename: string): TestSpec {
  const pos = { value: 0 };
  let name: string;

  if (peek(tokens, pos.value)?.type === 'TEST') {
    const testToken = consume(tokens, pos);
    name = testToken.value;
  } else {
    name = path.basename(filename, '.linespec');
  }

  const receiveToken = expectType(tokens, pos, 'RECEIVE');
  const httpReceiveMatch = receiveToken.value.match(/^HTTP:(\w+)\s+(.+)$/i);
  if (!httpReceiveMatch) {
    throw new LineSpecError(`Invalid RECEIVE format: ${receiveToken.value}`, receiveToken.line);
  }
  const method = httpReceiveMatch[1].toUpperCase();
  const pathValue = httpReceiveMatch[2];

  let receiveWithFile: string | undefined;
  if (peek(tokens, pos.value)?.type === 'WITH') {
    const withToken = consume(tokens, pos);
    receiveWithFile = withToken.value;
  }

  let receiveHeaders: Record<string, string> | undefined;
  if (peek(tokens, pos.value)?.type === 'HEADERS') {
    const headersToken = consume(tokens, pos);
    const headerLines = headersToken.value.split('\n').map(s => s.trim()).filter(s => s !== '');
    receiveHeaders = {};
    for (const line of headerLines) {
      const colonIndex = line.indexOf(':');
      if (colonIndex > 0) {
        const key = line.substring(0, colonIndex).trim();
        const value = line.substring(colonIndex + 1).trim();
        receiveHeaders[key] = value;
      }
    }
  }

  const receive: ReceiveStatement = {
    channel: 'HTTP',
    method,
    path: pathValue,
    withFile: receiveWithFile,
    headers: receiveHeaders,
  };

  const expects: ExpectStatement[] = [];

  while (peek(tokens, pos.value)?.type === 'EXPECT') {
    const expectToken = consume(tokens, pos);
    const expectPartial = parseExpectChannel(expectToken.value, expectToken.line) as ExpectStatement;

    let sql: string | undefined;
    if (peek(tokens, pos.value)?.type === 'USING_SQL') {
      const sqlToken = consume(tokens, pos);
      sql = sqlToken.value;
    }

    // Check for NO_TRANSACTION keyword
    let transactional = true; // Default
    if (peek(tokens, pos.value)?.type === 'NO_TRANSACTION') {
      consume(tokens, pos);
      transactional = false;
    }

    if (peek(tokens, pos.value)?.type === 'WITH') {
      const withToken = consume(tokens, pos);
      (expectPartial as any).withFile = withToken.value;
    }

    if (peek(tokens, pos.value)?.type === 'RETURNS') {
      const returnsToken = consume(tokens, pos);
      if (returnsToken.value === 'EMPTY') {
        (expectPartial as ExpectReadMysqlStatement).returnsEmpty = true;
      } else {
        (expectPartial as any).returnsFile = returnsToken.value;
      }
    }

    if (sql) {
      (expectPartial as any).sql = sql;
    }
    
    // Set transactional flag for WRITE_MYSQL
    if (expectPartial.channel === 'WRITE_MYSQL') {
      (expectPartial as ExpectWriteMysqlStatement).transactional = transactional;
    }

    // Parse VERIFY clauses
    const verifyRules: VerifyRule[] = [];
    while (peek(tokens, pos.value)?.type === 'VERIFY') {
      const verifyToken = consume(tokens, pos);
      const rule = parseVerifyRule(verifyToken.value, verifyToken.line);
      verifyRules.push(rule);
    }
    
    if (verifyRules.length > 0) {
      (expectPartial as any).verify = verifyRules;
    }

    expects.push(expectPartial);
  }

  const expectsNot: ExpectNotStatement[] = [];

  while (peek(tokens, pos.value)?.type === 'EXPECT_NOT') {
    const expectNotToken = consume(tokens, pos);
    const expectNotPartial = parseExpectNotChannel(expectNotToken.value, expectNotToken.line) as ExpectNotStatement;

    if (peek(tokens, pos.value)?.type === 'WITH') {
      const withToken = consume(tokens, pos);
      (expectNotPartial as any).withFile = withToken.value;
    }

    expectsNot.push(expectNotPartial);
  }

  const respondToken = expectType(tokens, pos, 'RESPOND');
  const httpRespondMatch = respondToken.value.match(/^HTTP:(\d+)$/i);
  if (!httpRespondMatch) {
    throw new LineSpecError(`Invalid RESPOND format: ${respondToken.value}`, respondToken.line);
  }
  const statusCode = parseInt(httpRespondMatch[1], 10);

  let respondWithFile: string | undefined;
  if (peek(tokens, pos.value)?.type === 'WITH') {
    const withToken = consume(tokens, pos);
    respondWithFile = withToken.value;
  }

  let respondNoise: string[] | undefined;
  if (peek(tokens, pos.value)?.type === 'NOISE') {
    const noiseToken = consume(tokens, pos);
    respondNoise = noiseToken.value.split('\n').map(s => s.trim()).filter(s => s !== '');
  }

  const respond: RespondStatement = {
    statusCode,
    withFile: respondWithFile,
    noise: respondNoise,
  };

  if (peek(tokens, pos.value) !== undefined) {
    const extraToken = peek(tokens, pos.value)!;
    throw new LineSpecError('No statements may appear after RESPOND', extraToken.line);
  }

  return {
    name,
    receive,
    expects,
    expectsNot,
    respond,
  };
}

File: pkg/dsl/parser.go
package dsl

import (
	"fmt"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/livecodelife/linespec/pkg/interpolate"
	"github.com/livecodelife/linespec/pkg/types"
)

// Package-level compiled regex patterns — compiled once at program startup.
var (
	// parseExpectStatement patterns
	reExpectHTTP = regexp.MustCompile(`^HTTP:(\w+)$`)
	reExpectMySQLWriteOp = regexp.MustCompile(`(?i)^(INSERT|UPDATE|DELETE)\s+(.+)$`)

	// parseVerifyRule patterns — query
	reVerifyQueryContains    = regexp.MustCompile(`(?i)^query\s+CONTAINS\s+['"](.+?)['"]$`)
	reVerifyQueryNotContains = regexp.MustCompile(`(?i)^query\s+NOT_CONTAINS\s+['"](.+?)['"]$`)
	reVerifyQueryMatches     = regexp.MustCompile(`(?i)^query\s+MATCHES\s/(.+?)/$`)

	// parseVerifyRule patterns — headers (HTTP and Kafka share same pattern)
	reVerifyHeadersContains    = regexp.MustCompile(`(?i)^headers\.([\w-]+)\s+CONTAINS\s+['"](.+?)['"]$`)
	reVerifyHeadersNotContains = regexp.MustCompile(`(?i)^headers\.([\w-]+)\s+NOT_CONTAINS\s+['"](.+?)['"]$`)
	reVerifyHeadersMatches     = regexp.MustCompile(`(?i)^headers\.([\w-]+)\s+MATCHES\s/(.+?)/$`)

	// parseVerifyRule patterns — body
	reVerifyBodyContains    = regexp.MustCompile(`(?i)^body\s+CONTAINS\s+['"](.+?)['"]$`)
	reVerifyBodyNotContains = regexp.MustCompile(`(?i)^body\s+NOT_CONTAINS\s+['"](.+?)['"]$`)
	reVerifyBodyMatches     = regexp.MustCompile(`(?i)^body\s+MATCHES\s/(.+?)/$`)

	// parseVerifyRule patterns — url
	reVerifyURLContains    = regexp.MustCompile(`(?i)^url\s+CONTAINS\s+['"](.+?)['"]$`)
	reVerifyURLNotContains = regexp.MustCompile(`(?i)^url\s+NOT_CONTAINS\s+['"](.+?)['"]$`)
	reVerifyURLMatches     = regexp.MustCompile(`(?i)^url\s+MATCHES\s/(.+?)/$`)

	// parseVerifyRule patterns — path
	reVerifyPathContains    = regexp.MustCompile(`(?i)^path\s+CONTAINS\s+['"](.+?)['"]$`)
	reVerifyPathNotContains = regexp.MustCompile(`(?i)^path\s+NOT_CONTAINS\s+['"](.+?)['"]$`)
	reVerifyPathMatches     = regexp.MustCompile(`(?i)^path\s+MATCHES\s/(.+?)/$`)

	// parseVerifyRule patterns — key (Kafka)
	reVerifyKeyContains    = regexp.MustCompile(`(?i)^key\s+CONTAINS\s+['"](.+?)['"]$`)
	reVerifyKeyNotContains = regexp.MustCompile(`(?i)^key\s+NOT_CONTAINS\s+['"](.+?)['"]$`)
	reVerifyKeyMatches     = regexp.MustCompile(`(?i)^key\s+MATCHES\s/(.+?)/$`)

	// parseVerifyRule patterns — value (Kafka)
	reVerifyValueContains    = regexp.MustCompile(`(?i)^value\s+CONTAINS\s+['"](.+?)['"]$`)
	reVerifyValueNotContains = regexp.MustCompile(`(?i)^value\s+NOT_CONTAINS\s+['"](.+?)['"]$`)
	reVerifyValueMatches     = regexp.MustCompile(`(?i)^value\s+MATCHES\s/(.+?)/$`)

	// parseVerifyRule patterns — request_body (gRPC)
	reVerifyRequestBodyContains    = regexp.MustCompile(`(?i)^request_body\s+CONTAINS\s+['"](.+?)['"]$`)
	reVerifyRequestBodyNotContains = regexp.MustCompile(`(?i)^request_body\s+NOT_CONTAINS\s+['"](.+?)['"]$`)
	reVerifyRequestBodyMatches     = regexp.MustCompile(`(?i)^request_body\s+MATCHES\s/(.+?)/$`)

	// parseVerifyRule patterns — metadata.NAME (gRPC)
	reVerifyMetadataContains    = regexp.MustCompile(`(?i)^metadata\.([\w-]+)\s+CONTAINS\s+['"](.+?)['"]$`)
	reVerifyMetadataNotContains = regexp.MustCompile(`(?i)^metadata\.([\w-]+)\s+NOT_CONTAINS\s+['"](.+?)['"]$`)
	reVerifyMetadataMatches     = regexp.MustCompile(`(?i)^metadata\.([\w-]+)\s+MATCHES\s/(.+?)/$`)

	// parseVerifyRule patterns — command (Redis)
	reVerifyCommandContains    = regexp.MustCompile(`(?i)^command\s+CONTAINS\s+['"](.+?)['"]$`)
	reVerifyCommandNotContains = regexp.MustCompile(`(?i)^command\s+NOT_CONTAINS\s+['"](.+?)['"]$`)
	reVerifyCommandMatches     = regexp.MustCompile(`(?i)^command\s+MATCHES\s/(.+?)/$`)

	// parseTestSpec / parseReceive patterns
	reReceiveHTTP    = regexp.MustCompile(`(?i)^HTTP:(\w+)\s+(.+)$`)
	reReceiveKafka   = regexp.MustCompile(`(?i)^(?:KAFKA|EVENT):(.+)$`)
	reReceiveGRPC    = regexp.MustCompile(`(?i)^GRPC:(.+)/(\w+)$`)
	reRespondStatus  = regexp.MustCompile(`(?i)^HTTP:(\d+)$`)
	reReturnsPayload = regexp.MustCompile(`^\{\{(.+)\}\}$`)

	// parseExpectChannel patterns
	reExpectEvent = regexp.MustCompile(`(?i)^(EVENT|MESSAGE):(.+)$`)
	reExpectGRPC  = regexp.MustCompile(`(?i)^GRPC:(.+)/(\w+)$`)
)

type Parser struct {
	tokens   []Token
	pos      int
	Resolver *interpolate.Resolver
}

func NewParser(tokens []Token) *Parser {
	return &Parser{tokens: tokens, pos: 0}
}

// NewParserWithResolver creates a parser with a resolver for variable substitution
func NewParserWithResolver(tokens []Token, resolver *interpolate.Resolver) *Parser {
	return &Parser{tokens: tokens, pos: 0, Resolver: resolver}
}

func (p *Parser) peek() *Token {
	if p.pos >= len(p.tokens) {
		return nil
	}
	return &p.tokens[p.pos]
}

func (p *Parser) consume() *Token {
	token := p.peek()
	if token != nil {
		p.pos++
	}
	return token
}

func (p *Parser) expect(tType TokenType) (*Token, error) {
	token := p.consume()
	if token == nil || token.Type != tType {
		msg := fmt.Sprintf("Expected %s but got %v", tType, token)
		if token != nil {
			return nil, fmt.Errorf("Parser error at line %d: %s", token.Line, msg)
		}
		return nil, fmt.Errorf("%s", msg)
	}
	return token, nil
}

func (p *Parser) Parse(filename string) (*types.TestSpec, error) {
	spec := &types.TestSpec{}
	spec.FilePath = filename
	spec.BaseDir = filepath.Dir(filename)

	if p.peek().Type == TokenTest {
		token := p.consume()
		spec.Name = token.Literal
	} else {
		spec.Name = strings.TrimSuffix(filepath.Base(filename), ".linespec")
	}

	receiveToken, err := p.expect(TokenReceive)
	if err != nil {
		return nil, err
	}

	if mHttp := reReceiveHTTP.FindStringSubmatch(receiveToken.Literal); mHttp != nil {
		spec.Receive.Channel = types.HTTP
		spec.Receive.Method = strings.ToUpper(mHttp[1])
		spec.Receive.Path = p.resolve(mHttp[2])
	} else if mKafka := reReceiveKafka.FindStringSubmatch(receiveToken.Literal); mKafka != nil {
		spec.Receive.Channel = types.Event
		spec.Receive.Topic = strings.TrimSpace(mKafka[1])
	} else if mGRPC := reReceiveGRPC.FindStringSubmatch(receiveToken.Literal); mGRPC != nil {
		spec.Receive.Channel = types.GRPC
		spec.Receive.Service = mGRPC[1]
		spec.Receive.RPCMethod = mGRPC[2]
	} else {
		return nil, fmt.Errorf("Invalid RECEIVE format at line %d: %s", receiveToken.Line, receiveToken.Literal)
	}

	if p.peek().Type == TokenWith {
		spec.Receive.WithFile = p.consume().Literal
	}

	if p.peek().Type == TokenHeaders {
		headersToken := p.consume()
		spec.Receive.Headers = p.resolveHeaders(parseHeaders(headersToken.Literal))
	}

	if p.peek().Type == TokenTimeout {
		timeoutToken := p.consume()
		d, err := time.ParseDuration(timeoutToken.Literal)
		if err != nil {
			return nil, fmt.Errorf("Invalid TIMEOUT value at line %d: %q (use Go duration syntax, e.g. 5m, 30s)", timeoutToken.Line, timeoutToken.Literal)
		}
		spec.Timeout = d
	}

	for p.peek().Type == TokenExpect {
		expect, err := p.parseExpect()
		if err != nil {
			return nil, err
		}
		spec.Expects = append(spec.Expects, *expect)
	}

	for p.peek().Type == TokenExpectNot {
		expectNot, err := p.parseExpectNot()
		if err != nil {
			return nil, err
		}
		spec.ExpectsNot = append(spec.ExpectsNot, *expectNot)
	}

	// RESPOND is required for HTTP-triggered tests, optional for Kafka consumer tests.
	if p.peek() != nil && p.peek().Type == TokenRespond {
		respondToken := p.consume()

		mStatus := reRespondStatus.FindStringSubmatch(respondToken.Literal)
		if mStatus == nil {
			return nil, fmt.Errorf("Invalid RESPOND format at line %d: %s", respondToken.Line, respondToken.Literal)
		}

		statusCode, _ := strconv.Atoi(mStatus[1])
		spec.Respond.StatusCode = statusCode

		if p.peek().Type == TokenWith {
			spec.Respond.WithFile = p.consume().Literal
		}

		if p.peek().Type == TokenNoise {
			noiseToken := p.consume()
			spec.Respond.Noise = strings.Split(strings.TrimSpace(noiseToken.Literal), "\n")
		}

		if p.peek().Type == TokenHeaders {
			headersToken := p.consume()
			spec.Respond.Headers = p.resolveHeaders(parseHeaders(headersToken.Literal))
		}
	} else if spec.Receive.Channel == types.HTTP {
		return nil, fmt.Errorf("RESPOND block is required for HTTP-triggered tests")
	}

	return spec, nil
}

// resolve applies variable substitution if a resolver is configured
func (p *Parser) resolve(s string) string {
	if p.Resolver != nil && interpolate.HasVariables(s) {
		return p.Resolver.Resolve(s)
	}
	return s
}

// resolveHeaders applies variable substitution to all header values
func (p *Parser) resolveHeaders(headers map[string]string) map[string]string {
	if p.Resolver == nil {
		return headers
	}
	return p.Resolver.ResolveMap(headers)
}

func (p *Parser) parseExpect() (*types.ExpectStatement, error) {
	token := p.consume()
	expect, err := parseExpectChannel(token.Literal, token.Line)
	if err != nil {
		return nil, err
	}

	// Apply variable substitution to channel-specific fields
	if expect.Channel == types.HTTP && expect.URL != "" {
		expect.URL = p.resolve(expect.URL)
	}
	if expect.Channel == types.ReadRedis || expect.Channel == types.WriteRedis {
		expect.RedisKey = p.resolve(expect.RedisKey)
	}
	if expect.Channel == types.GRPC {
		expect.Service = p.resolve(expect.Service)
		expect.RPCMethod = p.resolve(expect.RPCMethod)
	}

	// Handle HEADERS for HTTP expectations
	if p.peek().Type == TokenHeaders {
		headersToken := p.consume()
		expect.Headers = p.resolveHeaders(parseHeaders(headersToken.Literal))
	}

	if p.peek().Type == TokenUsingSql {
		p.consume() // TokenUsingSql
		sqlToken, err := p.expect(TokenSqlBlock)
		if err != nil {
			return nil, err
		}
		expect.SQL = p.resolve(sqlToken.Literal)
	}

	if p.peek().Type == TokenNoTransaction {
		p.consume()
		expect.NoTransaction = true
	}

	if p.peek().Type == TokenWith {
		expect.WithFile = p.consume().Literal
	}

	if p.peek().Type == TokenReturns {
		returnsToken := p.consume()
		if strings.ToUpper(returnsToken.Literal) == "EMPTY" {
			expect.ReturnsEmpty = true
		} else if m := reReturnsPayload.FindStringSubmatch(returnsToken.Literal); m != nil {
			expect.ReturnsFile = m[1]
		} else {
			expect.ReturnsFile = returnsToken.Literal
		}
	}

	if p.peek().Type == TokenResponseHeaders {
		headersToken := p.consume()
		expect.ResponseHeaders = p.resolveHeaders(parseHeaders(headersToken.Literal))
	}

	for p.peek().Type == TokenVerify {
		verifyToken := p.consume()
		rule, err := parseVerifyRule(verifyToken.Literal, verifyToken.Line)
		if err != nil {
			return nil, err
		}
		expect.Verify = append(expect.Verify, *rule)
	}

	return expect, nil
}

func (p *Parser) parseExpectNot() (*types.ExpectStatement, error) {
	token := p.consume()
	expect, err := parseExpectChannel(token.Literal, token.Line) // Using same logic for channel parsing
	if err != nil {
		return nil, err
	}
	expect.Negative = true

	if expect.Channel == types.ReadRedis || expect.Channel == types.WriteRedis {
		expect.RedisKey = p.resolve(expect.RedisKey)
	}
	if expect.Channel == types.GRPC {
		expect.Service = p.resolve(expect.Service)
		expect.RPCMethod = p.resolve(expect.RPCMethod)
	}

	if p.peek().Type == TokenWith {
		expect.WithFile = p.consume().Literal
	}

	return expect, nil
}

func parseHeaders(literal string) map[string]string {
	headers := make(map[string]string)
	lines := strings.Split(literal, "\n")
	for _, line := range lines {
		parts := strings.SplitN(line, ":", 2)
		if len(parts) == 2 {
			headers[strings.TrimSpace(parts[0])] = strings.TrimSpace(parts[1])
		}
	}
	return headers
}

func parseExpectChannel(value string, line int) (*types.ExpectStatement, error) {
	if m := reExpectEvent.FindStringSubmatch(value); m != nil {
		return &types.ExpectStatement{
			Channel: types.Event,
			Topic:   m[2],
		}, nil
	}

	// gRPC: GRPC:package.Service/Method (no space, matched before SplitN)
	if m := reExpectGRPC.FindStringSubmatch(value); m != nil {
		return &types.ExpectStatement{
			Channel:   types.GRPC,
			Service:   m[1],
			RPCMethod: m[2],
		}, nil
	}

	parts := strings.SplitN(value, " ", 2)
	if len(parts) < 2 {
		return nil, fmt.Errorf("Invalid EXPECT channel format at line %d: %s", line, value)
	}

	channelPart := strings.ToUpper(parts[0])
	rest := parts[1]

	if m := reExpectHTTP.FindStringSubmatch(channelPart); m != nil {
		return &types.ExpectStatement{
			Channel: types.HTTP,
			Method:  strings.ToUpper(m[1]),
			URL:     rest,
		}, nil
	}

	if channelPart == "WRITE:MYSQL" {
		if m := reExpectMySQLWriteOp.FindStringSubmatch(rest); m != nil {
			return &types.ExpectStatement{
				Channel: types.WriteMySQL,
				Table:   m[2],
			}, nil
		}
		return &types.ExpectStatement{
			Channel: types.WriteMySQL,
			Table:   rest,
		}, nil
	}

	if channelPart == "READ:MYSQL" {
		return &types.ExpectStatement{
			Channel: types.ReadMySQL,
			Table:   rest,
		}, nil
	}

	if channelPart == "WRITE:POSTGRESQL" {
		return &types.ExpectStatement{
			Channel: types.WritePostgreSQL,
			Table:   rest,
		}, nil
	}

	if channelPart == "READ:POSTGRESQL" {
		return &types.ExpectStatement{
			Channel: types.ReadPostgreSQL,
			Table:   rest,
		}, nil
	}

	// Redis: READ:REDIS <command> <key> or WRITE:REDIS <command> <key>
	if channelPart == "READ:REDIS" || channelPart == "WRITE:REDIS" {
		channel := types.ReadRedis
		if channelPart == "WRITE:REDIS" {
			channel = types.WriteRedis
		}
		redisParts := strings.SplitN(rest, " ", 2)
		cmd := strings.ToUpper(redisParts[0])
		key := ""
		if len(redisParts) > 1 {
			key = redisParts[1]
		}
		return &types.ExpectStatement{
			Channel:  channel,
			Command:  cmd,
			RedisKey: key,
		}, nil
	}

	return nil, fmt.Errorf("Unrecognized EXPECT channel at line %d: %s", line, channelPart)
}

func parseVerifyRule(value string, line int) (*types.VerifyRule, error) {
	// SQL targets
	if m := reVerifyQueryContains.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "CONTAINS", Target: "query", Pattern: m[1]}, nil
	}
	if m := reVerifyQueryNotContains.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "NOT_CONTAINS", Target: "query", Pattern: m[1]}, nil
	}
	if m := reVerifyQueryMatches.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "MATCHES", Target: "query", Pattern: m[1]}, nil
	}

	// HTTP targets: headers.NAME
	if m := reVerifyHeadersContains.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "CONTAINS", Target: "headers." + m[1], Pattern: m[2]}, nil
	}
	if m := reVerifyHeadersNotContains.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "NOT_CONTAINS", Target: "headers." + m[1], Pattern: m[2]}, nil
	}
	if m := reVerifyHeadersMatches.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "MATCHES", Target: "headers." + m[1], Pattern: m[2]}, nil
	}

	// HTTP targets: body
	if m := reVerifyBodyContains.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "CONTAINS", Target: "body", Pattern: m[1]}, nil
	}
	if m := reVerifyBodyNotContains.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "NOT_CONTAINS", Target: "body", Pattern: m[1]}, nil
	}
	if m := reVerifyBodyMatches.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "MATCHES", Target: "body", Pattern: m[1]}, nil
	}

	// HTTP targets: url
	if m := reVerifyURLContains.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "CONTAINS", Target: "url", Pattern: m[1]}, nil
	}
	if m := reVerifyURLNotContains.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "NOT_CONTAINS", Target: "url", Pattern: m[1]}, nil
	}
	if m := reVerifyURLMatches.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "MATCHES", Target: "url", Pattern: m[1]}, nil
	}

	// HTTP targets: path
	if m := reVerifyPathContains.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "CONTAINS", Target: "path", Pattern: m[1]}, nil
	}
	if m := reVerifyPathNotContains.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "NOT_CONTAINS", Target: "path", Pattern: m[1]}, nil
	}
	if m := reVerifyPathMatches.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "MATCHES", Target: "path", Pattern: m[1]}, nil
	}

	// Kafka targets: key
	if m := reVerifyKeyContains.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "CONTAINS", Target: "key", Pattern: m[1]}, nil
	}
	if m := reVerifyKeyNotContains.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "NOT_CONTAINS", Target: "key", Pattern: m[1]}, nil
	}
	if m := reVerifyKeyMatches.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "MATCHES", Target: "key", Pattern: m[1]}, nil
	}

	// Kafka targets: value
	if m := reVerifyValueContains.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "CONTAINS", Target: "value", Pattern: m[1]}, nil
	}
	if m := reVerifyValueNotContains.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "NOT_CONTAINS", Target: "value", Pattern: m[1]}, nil
	}
	if m := reVerifyValueMatches.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "MATCHES", Target: "value", Pattern: m[1]}, nil
	}

	// gRPC targets: request_body
	if m := reVerifyRequestBodyContains.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "CONTAINS", Target: "request_body", Pattern: m[1]}, nil
	}
	if m := reVerifyRequestBodyNotContains.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "NOT_CONTAINS", Target: "request_body", Pattern: m[1]}, nil
	}
	if m := reVerifyRequestBodyMatches.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "MATCHES", Target: "request_body", Pattern: m[1]}, nil
	}

	// gRPC targets: metadata.NAME
	if m := reVerifyMetadataContains.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "CONTAINS", Target: "metadata." + m[1], Pattern: m[2]}, nil
	}
	if m := reVerifyMetadataNotContains.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "NOT_CONTAINS", Target: "metadata." + m[1], Pattern: m[2]}, nil
	}
	if m := reVerifyMetadataMatches.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "MATCHES", Target: "metadata." + m[1], Pattern: m[2]}, nil
	}

	// Redis targets: command
	if m := reVerifyCommandContains.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "CONTAINS", Target: "command", Pattern: m[1]}, nil
	}
	if m := reVerifyCommandNotContains.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "NOT_CONTAINS", Target: "command", Pattern: m[1]}, nil
	}
	if m := reVerifyCommandMatches.FindStringSubmatch(value); m != nil {
		return &types.VerifyRule{Type: "MATCHES", Target: "command", Pattern: m[1]}, nil
	}

	return nil, fmt.Errorf("Invalid VERIFY format at line %d: %s", line, value)
}
